diff --git a/configure.ac b/configure.ac
index f58595b..bb39266 100644
--- a/configure.ac
+++ b/configure.ac
@@ -259,6 +259,9 @@ else
    AC_SUBST(M4RI_HAVE_LIBPNG)
 fi
 
+M4RI_BUILDING_M4RI=1
+AC_DEFINE([M4RI_BUILDING_M4RI], [1], [Define to indicate that m4ri is being built instead of being used])
+
 RELEASE="AC_PACKAGE_VERSION"
 AC_SUBST(RELEASE)
 
diff --git a/m4ri/graycode.h b/m4ri/graycode.h
index db86750..23cad7e 100644
--- a/m4ri/graycode.h
+++ b/m4ri/graycode.h
@@ -12,6 +12,22 @@
 #ifndef M4RI_GRAYFLEX_H
 #define M4RI_GRAYFLEX_H
 
+#ifdef HAVE_CONFIG_H
+ #include "config.h"
+#endif
+
+#ifdef _WIN32
+  #if defined(DLL_EXPORT) && defined(M4RI_BUILDING_M4RI)
+    #define M4RI_DLL_EXPORT __declspec(dllexport)
+  #elif defined(M4RI_USE_DLL) && !defined(M4RI_BUILDING_M4RI)
+    #define M4RI_DLL_EXPORT __declspec(dllimport)
+  #else
+    #define M4RI_DLL_EXPORT
+  #endif
+#else
+  #define M4RI_DLL_EXPORT
+#endif
+
 /******************************************************************************
 *
 *                 M4RI: Linear Algebra over GF(2)
@@ -79,7 +95,7 @@ typedef struct {
  * \warning Not thread safe!
  */ 
 
-extern code **m4ri_codebook;
+M4RI_DLL_EXPORT extern code **m4ri_codebook;
 
 /**
  * Returns the i-th gray code entry for a gray code of length \f$2^l\f$.
diff --git a/testsuite/test_solve.c b/testsuite/test_solve.c
index 023f124..80d8385 100644
--- a/testsuite/test_solve.c
+++ b/testsuite/test_solve.c
@@ -72,8 +72,8 @@ int main() {
   srandom(17);
 
   for(size_t i=0; i<100; i++) {
-    size_t m = random() & 511;
-    size_t n = random() & 1023;
+    size_t m = m4ri_random_word() & 511;
+    size_t n = m4ri_random_word() & 1023;
     m = m ? (m) : 1;
     n = n ? (n) : 1;
 
diff --git a/testsuite/testing.h b/testsuite/testing.h
index dece77a..b9021c9 100644
--- a/testsuite/testing.h
+++ b/testsuite/testing.h
@@ -4,6 +4,10 @@
 #include <m4ri/misc.h>
 #include <m4ri/mzd.h>
 
+#ifdef _MSC_VER
+#define srandom srand
+#endif
+
 mzd_t *mzd_init_test_matrix_random(rci_t M, rci_t N, rci_t m, rci_t n, word pattern, mzd_t **A);
 void mzd_free_test_matrix_random(mzd_t *A, mzd_t *a);
 int mzd_check_pattern(mzd_t *A, rci_t m, rci_t n, word pattern);
diff --git a/testsuite/test_elimination.c b/testsuite/test_elimination.c
index 77a2f1c..db80e29 100644
--- a/testsuite/test_elimination.c
+++ b/testsuite/test_elimination.c
@@ -1,6 +1,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 int elim_test_equality(rci_t nr, rci_t nc) {
   int ret = 0; 
diff --git a/testsuite/test_multiplication.c b/testsuite/test_multiplication.c
index 3826604..4b56eb0 100644
--- a/testsuite/test_multiplication.c
+++ b/testsuite/test_multiplication.c
@@ -1,6 +1,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 /**
  * Check that the results of all implemented multiplication algorithms

diff --git a/testsuite/test_invert.c b/testsuite/test_invert.c
index 065b752..866acec 100644
--- a/testsuite/test_invert.c
+++ b/testsuite/test_invert.c
@@ -1,6 +1,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 /**
  * Check that inversion works.
diff --git a/testsuite/test_kernel.c b/testsuite/test_kernel.c
index 2df7690..ed62ef6 100644
--- a/testsuite/test_kernel.c
+++ b/testsuite/test_kernel.c
@@ -21,6 +21,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 int test_kernel_left_pluq(rci_t m, rci_t n) {
   mzd_t* A = mzd_init(m, n);
diff --git a/testsuite/test_misc.c b/testsuite/test_misc.c
index f093f9b..a9a9547 100644
--- a/testsuite/test_misc.c
+++ b/testsuite/test_misc.c
@@ -21,6 +21,7 @@
 
 #include <stdarg.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 #define b(n) (m4ri_one<<(n))
 
diff --git a/testsuite/test_ple.c b/testsuite/test_ple.c
index 34b31b5..660fa76 100644
--- a/testsuite/test_ple.c
+++ b/testsuite/test_ple.c
@@ -1,6 +1,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 int check_pluq(mzd_t *A, rci_t *r) {
   mzd_t* Acopy = mzd_copy (NULL,A);
diff --git a/testsuite/test_random.c b/testsuite/test_random.c
index 5beea16..dc248ae 100644
--- a/testsuite/test_random.c
+++ b/testsuite/test_random.c
@@ -26,6 +26,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 int test_random(rci_t m, rci_t n)
 {
diff --git a/testsuite/test_smallops.c b/testsuite/test_smallops.c
index df493bb..beca1af 100644
--- a/testsuite/test_smallops.c
+++ b/testsuite/test_smallops.c
@@ -3,6 +3,7 @@
 #include <inttypes.h>
 #include "testing.h"
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 //#define ABORT_ON_FAIL 1
 
diff --git a/testsuite/test_trsm.c b/testsuite/test_trsm.c
index afa697b..ca540bd 100644
--- a/testsuite/test_trsm.c
+++ b/testsuite/test_trsm.c
@@ -1,6 +1,7 @@
 #include <m4ri/config.h>
 #include <stdlib.h>
 #include <m4ri/m4ri.h>
+#include "testing.h"
 
 int test_trsm_upper_right (rci_t m, rci_t n, int offset){
   printf("upper_right:: m: %4d n: %4d offset: %4d ... ", m, n, offset);
