commit cf985b3479a08b83dde7147c8cb47ea87c73e96e
Author: Isuru Fernando <isuruf@gmail.com>
Date:   Sun Sep 16 01:51:06 2018 -0500

    Don't include unistd if on MSVC

diff --git a/m4ri/djb.c b/m4ri/djb.c
index b111bc2..1020c06 100644
--- a/m4ri/djb.c
+++ b/m4ri/djb.c
@@ -13,7 +13,9 @@
 #include <m4ri/djb.h>
 #include <m4ri/xor.h>
 
+#ifndef _MSC_VER
 #include <unistd.h>
+#endif
 #include <stdlib.h>
 
 static inline int mzd_compare_rows_revlex(const mzd_t *A, rci_t a, rci_t b) {
